src/EOS.hpp

```diff
@@ -49,42 +59,111 @@ template <typename problem_t> class EOS
 
 template <typename problem_t>
 AMREX_FORCE_INLINE AMREX_GPU_HOST_DEVICE auto EOS<problem_t>::ComputeTgasFromEint(amrex::Real rho, amrex::Real Eint,
-										  std::optional<amrex::GpuArray<amrex::Real, nmscalars_>> massFractions)
+										  std::optional<amrex::GpuArray<amrex::Real, nmscalars_>> massScalars)
     -> amrex::Real
 {
 	// return temperature for an ideal gas
 	amrex::Real Tgas = NAN;
-	if constexpr (gamma_ != 1.0) {
-		const amrex::Real c_v = boltzmann_constant_ / (mean_molecular_weight_ * (gamma_ - 1.0));
-		Tgas = Eint / (rho * c_v);
+
+	if constexpr (PRIMORDIAL_CHEM_ENABLED) {
+		burn_t chemstate;
+		chemstate.rho = rho;
+		chemstate.e = Eint / rho;
+		// initialize array of number densities
+		for (int ii = 0; ii < NumSpec; ++ii) {
+			chemstate.xn[ii] = -1.0;
+		}
+
+		if (massScalars.has_value()) {
+			const auto &massArray = massScalars.value();
+			for (int nn = 0; nn < nmscalars_; ++nn) {
+				chemstate.xn[nn] = massArray[nn] / spmasses[nn]; // massScalars are partial densities (massFractions * rho)
+			}
+		}
+
+		eos(eos_input_re, chemstate);
+		amrex::Real Tgas = chemstate.T;
+	} else {
+		amrex::Real Tgas = NAN;
+		if constexpr (gamma_ != 1.0) {
+			const amrex::Real c_v = boltzmann_constant_ / (mean_molecular_weight_ * (gamma_ - 1.0));
+			Tgas = Eint / (rho * c_v);
+		}
 	}
 	return Tgas;
 }
 
 template <typename problem_t>
 AMREX_FORCE_INLINE AMREX_GPU_HOST_DEVICE auto EOS<problem_t>::ComputeEintFromTgas(amrex::Real rho, amrex::Real Tgas,
-										  std::optional<amrex::GpuArray<amrex::Real, nmscalars_>> massFractions)
+										  std::optional<amrex::GpuArray<amrex::Real, nmscalars_>> massScalars)
     -> amrex::Real
 {
 	// return internal energy density for a gamma-law ideal gas
 	amrex::Real Eint = NAN;
-	if constexpr (gamma_ != 1.0) {
-		const amrex::Real c_v = boltzmann_constant_ / (mean_molecular_weight_ * (gamma_ - 1.0));
-		Eint = rho * c_v * Tgas;
+
+	if constexpr (PRIMORDIAL_CHEM_ENABLED) {
+		burn_t chemstate;
+		chemstate.rho = rho;
+		// Define and initialize Tgas here
+		amrex::Real Tgas_value = Tgas;
+		chemstate.T = Tgas_value;
+		// initialize array of number densities
+		for (int ii = 0; ii < NumSpec; ++ii) {
+			chemstate.xn[ii] = -1.0;
+		}
+
+		if (massScalars.has_value()) {
+			const auto &massArray = massScalars.value();
+			for (int nn = 0; nn < nmscalars_; ++nn) {
+				chemstate.xn[nn] = massArray[nn] / spmasses[nn]; // massScalars are partial densities (massFractions * rho)
+			}
+		}
+
+		eos(eos_input_rt, chemstate);
+		amrex::Real const Eint = chemstate.e * chemstate.rho;
```

Comment:
This value of `Eint` shadows the `Eint` defined in the enclosing scope. So this function will always return `NAN`. You can fix this by changing this line to just `Eint = chemstate.e * chemstate.rho;`.